### ISSUE TITLE
access=destination not respected

### ISSUE THREAD

**Issue by nilsnolde** [valhalla/valhalla#2563]
I was surprised to see Valhalla (by default) doesn't deal with `access=destination` tags very well, i.e. using these ways (more or less) only when starting/ending the route on ways tagged that way. Often it's used with `highway=service` which also gets treated in a special way, so it's not super easy to find proper edges to test.

Finally I found a perfect edge with super simple tagging: https://www.openstreetmap.org/way/503170183.

I tried all FOSS routers and they all avoid that edge, e.g. routes for [OSRM/Graphhopper](https://www.openstreetmap.org/directions?engine=fossgis_osrm_car&route=48.08409%2C16.29337%3B48.08471%2C16.29569) and [ORS](https://maps.openrouteservice.org/directions?n1=48.083614&n2=16.294852&n3=18&a=48.0841,16.293438,48.084731,16.295659&b=0&c=0&k1=en-US&k2=km).

A fresh Vienna dataset with Valhalla master (and no changes to valhalla.json or costing models) goes right through it, same as the Mapbox server.

I assume Valhalla _should_ add a (fairly high) cost to those edges if the start/end point is not access=destination? Is there a config I overlooked?

**Comment by dnesbitt61** [comment]
My guess is Valhalla processes access::destination properly - assigning high cost when such a way is used outside the origin/destination of a route. This particular case applies only to motor_vehicle (motor_vehicle:destination tag) and I'm not sure that is handled. The destination flag (dest_only) on a DirectedEdge is applied to all modes of travel - not separated by particular modes - so this could be another factor in properly handling all such tags.

**Comment by nilsnolde** [comment]
Oh right, it's not access=destination.. I'll that tmrw though, makes sense. `motor_vehicle=destination` is quite common as well actually, 80k (300k for access=destination). Sounds like a bigger job to implement. Not sure if I'd find the time to look into it anytime soon, but I'll keep it in mind.

**Comment by nilsnolde** [comment]
Quick follow-up: access=destination works. As you say, for all vehicles. Not sure about other countries, but in Germany in reality `destination` tags are most often targeted at vehicles (pretty sure a lot of `access=destination` should be `motor_vehicle=destination` here), the "Anlieger frei", and free for bike/ped. IMHO it's not the tiniest flaw to have a mode-independent destination flag. Or am I super biased and other countries handle it differently?

**Comment by dnesbitt61** [comment]
I also think it is fine to have a generic destination flag. Each costing method has a configurable penalty (destination_only_penalty) so users can decide how much penalty to apply for each mode of travel. So I guess this issue boils down to supporting motor_vehicle=destination tags - having them set the destination flag on directed edges when this tag is applied.

**Comment by nilsnolde** [comment]
Ah, didn't know about `destination_only_penalty`.. Guess that never made it into the API docs. 

Still, if we were to set the destination flag for motor_vehicle=destination, by default also pedestrian/bikes would be affected. To remedy, setting destination_only_penalty for those profiles to 0, they would also use access=destination roads as thru roads, which is probably also not super desirable. I think in that case it might actually be better to not set a destination flag for motor_vehicle=destination. With generic destination flags it's probably less wrong to also only consider generic destination tags, rather than generalizing specific tags. 

Would it be a lot of work (or even accepted) to make destination flags somehow work profile-dependent?

**Comment by nilsnolde** [comment]
To update on this, since we have this use case very soon with a client: when routing trucks, we'd ideally have Valhalla respect truck/mode-specific `destination_only` so we can have edges which a truck is not allowed on unless it's off-loading but a car can traverse without penalties. It's a very common issue, Valhalla is perfectly suited for truck routing, so I think it'd be really good to support that. 

AFAICT, with the current logic, we'll either have to set `destination_only_penalty = 600` (minimum) for car requests while flagging all edges as `dest_only` edges which are only truck-specific but don't apply to cars, and accept that car routing will be much less likely to take an edge which it is actually allowed on, and also doesn't give fine-grained control for which mode to set which `dest_only` penalty. Or alternatively, we don't flag truck-specific `dest_only` edges and accept that trucks can take edges which they're actually not allowed on (not really acceptable).

I'm not sure though if there need to be other mode-specific tags, so it might make sense to only have an additional `truck_dest_only` property instead of support for all modes (in spirit the same as the specific `truck_speed`, so there's a much costlier precedence;)). Then we'd have 2 `dest_only` properties on the edges, one for general/car which applies to **all** modes of transport (incl. truck) and one specific `truck_dest_only` which is only respected by hgv. Without looking into the code: sounds like a single boolean to DE & labels to me? So, just 
- bit of work in the lua to support `hgv = destination` and friends
- parse that tag and add it to the directed edges
- a few lines in the truck costing to look at `truck_dest_only` unless it already has a generic `dest_only` (no double penalty)

And we pay one bit (DE would have 4 after that) and a few lines of code. Would sound fair to me.

@dnesbitt61 @kevinkreiser anything comes to mind that makes the idea unreasonable? Or maybe I'm not making sense/see it too complicated? Otherwise I'll do a deep dive in a while.

**Comment by dnesbitt61** [comment]
A single bit for truck_dest_only makes sense to me. Other modes should be able to use the regular dest_only flag. Bicycle perhaps could use a separate flag. We added an _ignore_dest_only_for_bicycles_ flag that when present will ignore destination only for bicycle routes (no penalty added).

**Comment by nilsnolde** [comment]
> We added an ignore_dest_only_for_bicycles flag that when present will ignore destination only for bicycle routes (no penalty added).

@dnesbitt61 You mean in your fork I assume? Is there OSM tagging for bike exceptions of dest_only? Just curious about the kind of ways that do restrict destination_only for bike/ped. I kind of figured bike/ped is mostly irrelevant for dest_only..

**Comment by dnesbitt61** [comment]
Yea, this is a fork of Valhalla. We apply a modest penalty to bicycles for destination only roads but there are times we want to not apply the penalty. I think if the way is part of a bicycle network we ignore the destination only penalty - there may be other cases where it is set as well to avoid adding the penalty.

**Comment by mangerlahn** [comment]
Valhalla seams to strictly avoid streets tagged with `motor_vehicle=destination` for bicycle routing, even if the given street is a designated bike street. In this case, the "Härtelstraße" is a bike street, but avoided by routing.

<img width="847" alt="Screenshot 2023-02-05 at 12 33 56" src="https://user-images.githubusercontent.com/11056766/216816422-c78a8e96-9351-4ba9-ba4d-5eb4c3597b22.png">
<img width="830" alt="Screenshot 2023-02-05 at 12 33 49" src="https://user-images.githubusercontent.com/11056766/216816427-48744312-f30c-488e-8de8-857a0e231f84.png">
https://www.openstreetmap.org/directions?engine=fossgis_valhalla_bicycle&route=51.33278%2C12.37365%3B51.33271%2C12.37739#map=18/51.33310/12.37539

**Comment by nilsnolde** [comment]
Yeah, that's a known problem, a variant of what @dnesbitt61 was talking about (one of their cases was a bike network relation). Valhalla doesn't do vehicle-specific destonly attributes. We're quickly running out of bits in the relevant structure, though I do see how in this case it's really bad (being an avid bike commuter I can relate how bad..). 

If we spared another bit for bicycle-specific destonly/opt-out-of-destonly (and the already agreed one for truck-specific destonly), we'd be down to 2 in directededge.h :fearful: Still, both are quite necessary IMO.

**Comment by mangerlahn** [comment]
Relates to #4081 and #4230.

**Comment by chrneumann** [comment]
> If we spared another bit for bicycle-specific destonly/opt-out-of-destonly (and the already agreed one for truck-specific destonly), we'd be down to 2 in directededge.h 😨 Still, both are quite necessary IMO.

Just stumbled onto the problem for bike routes (building an web app that uses Valhalla for routing). In my area, this gives quite some bad routes. What is the problem with adding more bits to the structure? Backward compatibility (reusing tiles from previous versions?)? Or is it not easy to add more bits in the first place / changing the code around this?

**Comment by chrstnbwnkl** [comment]
Adding bits is not difficult, the issue is backwards compatibility. Simply slapping another 64 bits to DirectedEdge and calling it v4 is not the favorable path forward. @kevinkreiser has talked about what he would like to do for v4, I remember something about grouping edge attributes by mode of transport and make those groups optional depending on which transport modes you actually care about, but maybe he can elaborate if there isn't a write-up of this somewhere already :smile:  The gist is, the amount of work needed to make this happen is quite large.

**Comment by chrneumann** [comment]
> the issue is backwards compatibility

But only because old tiles can't be reused? Just trying to understand the problem (if the problem is about old tiles): Are there use cases where tiles get that old? I mean, you possibly don't want to recreate tiles for every bug fix / most minor versions, but at the same time, shouldn't most of the Valhalla instances use tiles that are recreated regularly, so that the backward compatibility isn't such a great issue?

**Comment by dnesbitt61** [comment]
Most server based systems will refresh their tiles frequently. However, mobile/embedded Valhalla uses often download and store tiles for longer periods of time. It can be costly and burdensome for users to be forced to update data frequently.
